Upgrades: use the management address type from old install

gthvn1 opened the pull request.

gthvn1 committed:
Use the management address type from old install

During an upgrade we need to use the management address from the installation
read from /etc/xensource-inventory.  This does not change the behaviour of
installation.

Signed-off-by: Guillaume <guillaume.thouvenin@vates.tech>
Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

benjamreis commented on util.py:
```diff
@@ -357,6 +357,9 @@ def udevinfoCmd():
     return "".join([random.choice(string.ascii_lowercase) for x in range(6)])
 
+def mgmtAddrType():
+    return "Ipv4"
```
```suggestion
    return "IPv4"
```

gthvn1 commented:
It is really strange. When doing a fresh install using IPv6 static only my installation ends with IPv4. And the `/etc/xensource-inventory` has:
```
MANAGEMENT_INTERFACE=''
MANAGEMENT_ADDRESS_TYPE='IPv4'
```
I will add some logs to understand why it doesn't work...

gthvn1 commented on backend.py:
```diff
@@ -1659,7 +1660,9 @@ def writeInventory(installID, controlID, mounts, primary_disk, backup_partnum, s
     inv.write("MANAGEMENT_INTERFACE='%s'\n" % admin_bridge)
     # Default to IPv4 unless we have only got an IPv6 admin interface
-    if ((not admin_config.mode) and admin_config.modev6):
+    if mgmtAddrType:
```
we should probably use mgmtAddrType only if mode and modev6 are set to None

gthvn1 commented:
> It is really strange. When doing a fresh install using IPv6 static only my installation ends with IPv4. And the `/etc/xensource-inventory` has:
> 
> ```
> MANAGEMENT_INTERFACE=''
> MANAGEMENT_ADDRESS_TYPE='IPv4'
> ```
> 
> I will add some logs to understand why it doesn't work...

So the issue was as expected because the management address type is always set. By default it is set to IPv4. So we use it only when no modes are selected (and it is the case during an upgrade for example). I tested a fresh install and in this case we are either in IPv4 mode or IPv6 mode. If the mode is IPv6 only we are in IPv6 and if it is set to both we use IPv4. And as said during an upgrade we are using the one set in /etc/xensource-inventory.
I tested:
- fresh install in IPv6 (static and autoconf)
- fresh install in IPv4 (dhcp)
- upgrade from IPv6 static

ydirson commented on backend.py:
```diff
@@ -1658,8 +1659,11 @@ def writeInventory(installID, controlID, mounts, primary_disk, backup_partnum, s
     inv.write("DOM0_VCPUS='%d'\n" % host_config['dom0-vcpus'])
     inv.write("MANAGEMENT_INTERFACE='%s'\n" % admin_bridge)
+    # if no modes are configured then use mgmtAddrType
+    if not (admin_config.mode or admin_config.modev6):
+        inv.write("MANAGEMENT_ADDRESS_TYPE='%s'\n" % mgmtAddrType)
     # Default to IPv4 unless we have only got an IPv6 admin interface
-    if ((not admin_config.mode) and admin_config.modev6):
+    elif (not admin_config.mode) and admin_config.modev6:
```
wouldn't it be more readable to have a single `if not admin_config.mode` around all of this?

gthvn1 replied:
I think that cases are more readable like this but I can modify it no worries.

ydirson replied:
maybe with the updated logic we can still use `elif` to get get things more readable?

ydirson reviewed (changes requested):
The logic is not obvious at first read, it would be useful to give a few details about the change in the commit message

ydirson reviewed (commented):
Not really sure to get it:
- `mgmtAddrType()` seems to be used to set a default value, but only for install (would be useful to make it explicit in the code)
- `prepareUpgrade()` states `MANAGEMENT_ADDRESS_TYPE` must be present in `xensource-inventory` , but looks at the default value (as well as the uuids!?), even though exceptions are rised if they cannot get overriden by `getInventoryValue()` calls?

stormi commented:
Unless I'm wrong, I believe this PR is waiting for @gthvn1 after Yann's comments.

gthvn1 commented:
> Unless I'm wrong, I believe this PR is waiting for @gthvn1 after Yann's comments.

Yes right I missed it. But I'm not sure to understand the remark. As far as I understand it we will always in the case where no mode is selected we must have a value in the inventory from previous install so it will be overwritten with the value found in the inventory. And as the default case is "IPv4" I set it by default but yes maybe it is not needed. I don't remember exactly but I think when I tested it it didn't work without this definition.

gthvn1 commented:
I will give it a try and keep you in touch. Probably today.

gthvn1 commented:
> I will give it a try and keep you in touch. Probably today.

So when removing `def mgmtAddrType():` from util.py I have an error because "module" object has no attribute. So I added it. But I should probably try to follow what is done for "controlID".

gthvn1 commented:
@ydirson Ok I found why I added the `def mgmtAddrType`. It is because of ` Task(util.mgmtAddrType, As(ans), ['management-address-type']),`. And I did that because 'management-address-type' was added to `Task(writeInventory...` So I followed what is done for 'control-domain-uuid' but as I cannot use the same `Task(util.getUUID, As(ans), ['control-domain-uuid']),` I created one for management address...

gthvn1 committed:
Remove unused parameters passed to prepareUpgrade

The two parameters are read from the xensource inventory so no need to
pass them as parameters of the prepareUpgrade function.

Signed-off-by: Guillaume <guillaume.thouvenin@vates.tech>

gthvn1 commented:
I have splitted the PR in two commits. The first commit can be upstreamed.

gthvn1 commented:
- [x]  testing upgrade from IPv6 only
- [x]  testing upgrade from IPv4 only
- [x] test a fresh install with IPv6
    - Only tested using autoconf. Tried with dhcp but as our dhc6p server is not setup properly I'm not sure if the error is caused by our setup... 
- [x] test a fresh install with IPv4

stormi commented:
> I have split the PR in two commits. The first commit can be upstreamed.

The second commit can also be upstreamed as part of Benjamin's IPv6 PR, can't it?

gthvn1 commented:
Ah yes maybe :+1:

gthvn1 commented:
> > I have split the PR in two commits. The first commit can be upstreamed.
> 
> The second commit can also be upstreamed as part of Benjamin's IPv6 PR, can't it?

You think about this commit #17

stormi commented:
> > > I have split the PR in two commits. The first commit can be upstreamed.
> > 
> > 
> > The second commit can also be upstreamed as part of Benjamin's IPv6 PR, can't it?
> 
> You think about this commit #17

I don't know about that draft. I'm rather thinking of https://github.com/xenserver/host-installer/pull/2, if I'm not mistaken.

ydirson commented:
Rebased for v10.10.29

ydirson reviewed (changes requested):
There's a bug here, no parameter should be added to `prepareUpgrade`

ydirson reviewed (approved):
tested in automated install/upgrade/restore
